src/checkers/inference/solver/backend/z3/Z3BitVectorFormatTranslator.java

```diff
@@ -7,56 +7,42 @@
 import javax.annotation.processing.ProcessingEnvironment;
 import javax.lang.model.element.AnnotationMirror;
 
+import checkers.inference.solver.backend.AbstractFormatTranslator;
+import checkers.inference.solver.backend.z3.encoder.Z3BitVectorEqualityConstraintEncoder;
+import checkers.inference.solver.backend.z3.encoder.Z3BitVectorPreferenceConstraintEncoder;
+import checkers.inference.solver.backend.z3.encoder.Z3BitVectorSubtypeConstraintEncoder;
+import checkers.inference.util.ConstraintVerifier;
 import com.microsoft.z3.BitVecExpr;
 import com.microsoft.z3.BitVecNum;
 import com.microsoft.z3.BoolExpr;
 import com.microsoft.z3.Context;
 import com.microsoft.z3.Optimize;
 
 import checkers.inference.model.CombVariableSlot;
-import checkers.inference.model.CombineConstraint;
-import checkers.inference.model.ComparableConstraint;
 import checkers.inference.model.ConstantSlot;
-import checkers.inference.model.EqualityConstraint;
-import checkers.inference.model.ExistentialConstraint;
 import checkers.inference.model.ExistentialVariableSlot;
-import checkers.inference.model.InequalityConstraint;
-import checkers.inference.model.PreferenceConstraint;
 import checkers.inference.model.RefinementVariableSlot;
-import checkers.inference.model.Slot;
-import checkers.inference.model.SubtypeConstraint;
 import checkers.inference.model.VariableSlot;
-import checkers.inference.solver.backend.FormatTranslator;
 import checkers.inference.solver.frontend.Lattice;
-import checkers.inference.solver.frontend.VariableCombos;
 
-public abstract class Z3BitVectorFormatTranslator implements FormatTranslator<BitVecExpr, BoolExpr, BitVecNum>{
+public abstract class Z3BitVectorFormatTranslator extends AbstractFormatTranslator<BitVecExpr, BoolExpr, BitVecNum> {
 
     private Optimize solver;
 
     private Map<Integer, BitVecExpr> serializedSlots;
 
     protected Context context;
 
-    protected final BoolExpr EMPTY_VALUE = null;
-
-    protected final Lattice lattice;
-
     protected final Z3BitVectorCodec z3BitVectorCodec;
 
-    protected Z3SubtypeVariableCombos subtypeVariableCombos;
-
-    protected Z3EqualityVariableCombos equalityVariableCombos;
-
-    public Z3BitVectorFormatTranslator(Lattice lattice) {
-        this.lattice = lattice;
+    public Z3BitVectorFormatTranslator(Lattice lattice, ConstraintVerifier verifier) {
+        super(lattice, verifier);
         z3BitVectorCodec = createZ3BitVectorCodec();
         serializedSlots = new HashMap<>();
-        subtypeVariableCombos = new Z3SubtypeVariableCombos(EMPTY_VALUE);
-        equalityVariableCombos = new Z3EqualityVariableCombos(EMPTY_VALUE);
     }
 
-    protected boolean isSubtypeSubSet() {
+    // TODO Charles, can this be public? Must it stay in Z3BitVectorFormatTranslator?
+    public boolean isSubtypeSubSet() {
```

Comment:
I think this `isSubtypeSubSet` is only used in encoding subtype constraint. So maybe we can move it into Z3SubtypeConstraintEncoder. Let me try to do this.

Reply:
Sure, feel free to refactor any Z3 specific staff if you think it's proper.